### ISSUE TITLE
R-Package in new R version 3.5.0

### ISSUE THREAD

**Issue by weltherrschaf** [Microsoft/LightGBM#1341]
Is the R-Package of LGB supported under the latest R release? Can't get it to work yet, is anybody experiencing similar issues?

## Environment info
Operating System: Windows 10
C++/Python/R version: R: 3.5.0

## Error Message:
Just updating R and using the old LGB install I get:

"Error: package or namespace load failed for ‘lightgbm’:
 package ‘lightgbm’ was installed by an R version with different internals; it needs to be reinstalled for use with this R version"

After reinstalling LGB using the visual studio approach without errors, I'm getting this error when trying to run a minimal example:

"Loading required package: Matrix
Error in data$construct() : 
  lgb.Dataset.construct: cannot create Dataset handle"

## Reproducible examples

library(lightgbm)
data(agaricus.train, package = "lightgbm")
train <- agaricus.train
dtrain <- lgb.Dataset(train$data, label = train$label)
params <- list(objective = "regression", metric = "l2")
model <- lgb.cv(params,
                dtrain,
                10,
                nfold = 5,
                min_data = 1,
                learning_rate = 1,
                early_stopping_rounds = 10)

## Steps to reproduce

1. Start with a Running LGB install on R 3.4.3
2. Update R to either R 3.5.0 or R 3.5.0patched
3. Build new LGB

I tried several clean installs, getting the same error though

**Comment by guolinke** [comment]
@Laurae2

**Comment by Laurae2** [comment]
@guolinke I am getting the following in `construct`:

```r
R version 3.5.0 (2018-04-23) -- "Joy in Playing"
Copyright (C) 2018 The R Foundation for Statistical Computing
Platform: x86_64-w64-mingw32/x64 (64-bit)

R is free software and comes with ABSOLUTELY NO WARRANTY.
You are welcome to redistribute it under certain conditions.
Type 'license()' or 'licence()' for distribution details.

R is a collaborative project with many contributors.
Type 'contributors()' for more information and
'citation()' on how to cite R or R packages in publications.

Type 'demo()' for some demos, 'help()' for on-line help, or
'help.start()' for an HTML browser interface to help.
Type 'q()' to quit R.

(...)

Browse[2]> lgb.call("LGBM_DatasetCreateFromCSC_R", 
+          ret = handle, private$raw_data@p, private$raw_data@i, 
+          private$raw_data@x, length(private$raw_data@p), 
+          length(private$raw_data@x), nrow(private$raw_data), 
+          params_str, ref_handle)
[1] NA
```

I believe this is related to the following: https://svn.r-project.org/R/branches/ALTREP/ALTREP.html - it changed the memory layout of R objects.

Objects:

```r
Browse[2]> str(private$raw_data)
Formal class 'dgCMatrix' [package "Matrix"] with 6 slots
  ..@ i       : int [1:143286] 2 6 8 11 18 20 21 24 28 32 ...
  ..@ p       : int [1:127] 0 369 372 3306 5845 6489 6513 8380 8384 10991 ...
  ..@ Dim     : int [1:2] 6513 126
  ..@ Dimnames:List of 2
  .. ..$ : NULL
  .. ..$ : chr [1:126] "cap-shape=bell" "cap-shape=conical" "cap-shape=convex" "cap-shape=flat" ...
  ..@ x       : num [1:143286] 1 1 1 1 1 1 1 1 1 1 ...
  ..@ factors : list()

Browse[2]> params_str
 [1] 6f 62 6a 65 63 74 69 76 65 3d 72 65 67 72 65 73 73 69 6f 6e 20 6d 65 74 72 69 63 3d 6c 32 20 6d 69 6e 5f 64 61 74 61 3d 31 20 6c 65 61 72 6e 69 6e 67 5f 72 61
[54] 74 65 3d 31 20 76 65 72 62 6f 73 65 3d 31 00

Browse[2]> ref_handle
NULL
```

**Comment by guolinke** [comment]
it seems a big changes. However, I don’t have time to fix it recently. 
@Laurae2 Can you help to add the warnings to R package until fix it?

**Comment by Laurae2** [comment]
@guolinke I'll prevent R installation under >=3.5.0 until fixed.

**Comment by guolinke** [comment]
fixed in the latest commit
